In [ ]:
import pandas as pd

In [ ]:
from google.colab import files

In [ ]:
df = pd.read_csv(r'/content/topics_sups.csv', encoding='latin-1') #reading the given data set

In [ ]:
df.head()
# head() used to display first five tupels in a data set

ID                                               Name  \
0   1                     Air and Space Vehicles (5) NSc   
1   2  Special Topics in Aeronautics and Astronautics...   
2   3                        Engineering Statics (4) NSc   
3   4                             Thermodynamics (4) NSc   
4   5              Undergraduate Research (1-5, max. 10)   

                             cleaned_text  \
0                   r n pce Vehcle (5) Nc   
1  pecl pc n ernuc n rnuc (1-5, x. 10) Nc   
2                      Engneerng c (4) Nc   
3                            hernc (4) Nc   
4            Unergrue Reerch (1-5, x. 10)   

                              lemmatized_text  Cluster  
0                     r n pce Vehcle ( 5 ) Nc        0  
1  pecl pc n ernuc n rnuc ( 1-5 , x . 10 ) Nc        2  
2                        Engneerng c ( 4 ) Nc        0  
3                              hernc ( 4 ) Nc        0  
4            Unergrue Reerch ( 1-5 , x . 10 )        1

# CHECK FOR NULL VALUES IN THE DATA SET

In [ ]:
df.isnull().sum()
#isnull is used for checking whether a value is null or not
# sum() method is used for counting number of tuples

ID        0
Name    209
dtype: int64

We have found that we have 209 null values in the column 'Name'

Fill the null values with mean values

In [ ]:
len(df) #finding length of data set

326728

In [ ]:
df.dropna(inplace=True) # removing the null values

In [ ]:
len(df) # length of data set after removing the null values

326519

Finding the duplicate vales in data set

In [ ]:
d_n = df[df.duplicated(keep='first')]
#duplicated is used for finding duplicate values
#first will ensure that first value is preserved

In [ ]:
d_n

Empty DataFrame
Columns: [ID, Name]
Index: []

We found no duplicate values

Removing the stopwords and cleaning the columns in the data set


In [ ]:
from nltk.corpus import stopwords
import nltk

In [ ]:
nltk.download('stopwords')#downloading the stopwords

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [ ]:
stop_words=set(stopwords.words('english'))#sorting the stopwords into the set to access it directly

In [ ]:
df['cleaned_text'] = df['Name'].apply(lambda tokens: ''.join([word for word in tokens if word.lower() not in stop_words]))#removing the stopwords and adding it in another column named 'cleaned_text'

In [ ]:
df['cleaned_text'].head()

0                     r n pce Vehcle (5) Nc
1    pecl pc n ernuc n rnuc (1-5, x. 10) Nc
2                        Engneerng c (4) Nc
3                              hernc (4) Nc
4              Unergrue Reerch (1-5, x. 10)
Name: cleaned_text, dtype: object

In [ ]:
from nltk.stem import WordNetLemmatizer
# using lemmitization to get root words

In [ ]:
nltk.download('wordnet')
nltk.download('punkt')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [ ]:
lemmatizer = WordNetLemmatizer()

In [ ]:
def lemmatize_text(text):
    words = nltk.word_tokenize(text)  # Tokenize the text into words
    lemmatized_words = [lemmatizer.lemmatize(word) for word in words]  # Lemmatize each word
    return ' '.join(lemmatized_words)  # Join the lemmatized words back into a sentence

In [ ]:
# Assuming 'cleaned_text' is the name of the column you want to lemmatize
df['lemmatized_text'] = df['cleaned_text'].apply(lemmatize_text)


In [ ]:
df.head()

ID                                               Name  \
0   1                     Air and Space Vehicles (5) NSc   
1   2  Special Topics in Aeronautics and Astronautics...   
2   3                        Engineering Statics (4) NSc   
3   4                             Thermodynamics (4) NSc   
4   5              Undergraduate Research (1-5, max. 10)   

                             cleaned_text  \
0                   r n pce Vehcle (5) Nc   
1  pecl pc n ernuc n rnuc (1-5, x. 10) Nc   
2                      Engneerng c (4) Nc   
3                            hernc (4) Nc   
4            Unergrue Reerch (1-5, x. 10)   

                              lemmatized_text  
0                     r n pce Vehcle ( 5 ) Nc  
1  pecl pc n ernuc n rnuc ( 1-5 , x . 10 ) Nc  
2                        Engneerng c ( 4 ) Nc  
3                              hernc ( 4 ) Nc  
4            Unergrue Reerch ( 1-5 , x . 10 )

In [ ]:
from sklearn.cluster import KMeans
#importing kmeans for clustering

In [ ]:
kmeans = KMeans(n_clusters=3, init='k-means++', max_iter=300, n_init=10, random_state=0)
#n_clusters is used to denote number of clusters
#init Initialization method for K-Means clustering, setting initial cluster centroids.
#`max_iter`  the max iterations for K-Means clustering convergence.
# `n_init` in K-Means: Number of times to reinitialize centroids.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
# importing tfidfvectorizer for converting categorical data to numerical data

In [ ]:
tfidf_vectorizer = TfidfVectorizer()


In [ ]:
tfidf_matrix = tfidf_vectorizer.fit_transform(df['cleaned_text'])
# converting cleaned_text data to numerical data

In [ ]:
tfidf_matrix

<326519x13296 sparse matrix of type '<class 'numpy.float64'>'
	with 1136816 stored elements in Compressed Sparse Row format>

In [ ]:
kmeans.fit(tfidf_matrix)

/usr/local/lib/python3.10/dist-packages/sklearn/cluster/_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3, random_state=0)

In [ ]:
cluster_labels = kmeans.labels_


In [ ]:
cluster_labels

array([0, 2, 0, ..., 0, 2, 1], dtype=int32)

In [ ]:
df['Cluster'] = cluster_labels


In [ ]:
df.head()

ID                                               Name  \
0   1                     Air and Space Vehicles (5) NSc   
1   2  Special Topics in Aeronautics and Astronautics...   
2   3                        Engineering Statics (4) NSc   
3   4                             Thermodynamics (4) NSc   
4   5              Undergraduate Research (1-5, max. 10)   

                             cleaned_text  \
0                   r n pce Vehcle (5) Nc   
1  pecl pc n ernuc n rnuc (1-5, x. 10) Nc   
2                      Engneerng c (4) Nc   
3                            hernc (4) Nc   
4            Unergrue Reerch (1-5, x. 10)   

                              lemmatized_text  Cluster  
0                     r n pce Vehcle ( 5 ) Nc        0  
1  pecl pc n ernuc n rnuc ( 1-5 , x . 10 ) Nc        2  
2                        Engneerng c ( 4 ) Nc        0  
3                              hernc ( 4 ) Nc        0  
4            Unergrue Reerch ( 1-5 , x . 10 )        1

In [ ]:
df.drop('cleaned_text',axis=1)

ID                                               Name  \
0            1                     Air and Space Vehicles (5) NSc   
1            2  Special Topics in Aeronautics and Astronautics...   
2            3                        Engineering Statics (4) NSc   
3            4                             Thermodynamics (4) NSc   
4            5              Undergraduate Research (1-5, max. 10)   
...        ...                                                ...   
326723  326724                  Current Issues in Feminist Theory   
326724  326725       Special Topics in Feminist Theory & Research   
326725  326726                    Cultural Study of Masculinities   
326726  326727                Special Study for Graduate Students   
326727  326728                    Dissertation Research & Writing   

                                   lemmatized_text  Cluster  
0                          r n pce Vehcle ( 5 ) Nc        0  
1       pecl pc n ernuc n rnuc ( 1-5 , x . 10 ) Nc        2  
2                             Engneerng c ( 4 ) Nc        0  
3                                   hernc ( 4 ) Nc        0  
4                 Unergrue Reerch ( 1-5 , x . 10 )        1  
...                                            ...      ...  
326723                         Curren ue n Fen her        0  
326724                  pecl pc n Fen her & Reerch        2  
326725                            Culurl u f culne        0  
326726                          pecl u fr Grue uen        2  
326727                           ern Reerch & Wrng        1  

[326519 rows x 4 columns]